### PR TITLE
Allocation domain support for fusion inputs

### PR DESCRIPTION
**Pull request by zasdfgbnm**
This PR is stacked on https://github.com/NVIDIA/Fuser/pull/337, please review that PR first.

This PR adds support for allocation domains in tensor inputs. The ATen tensor provided to the executor is always in the format of rFactor domain, and we need to analyze the relationship between rFactor domain and allocation domain to convert the strides in the format of allocation domain. The shape is still left in rFactor domain, because things like `T1.size[1]` in generated code comes from replacing the rFactor domain's extents with named scalar, so these should be in rFactor domain.

### CODE REVIEW

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
!build

**Comment by zasdfgbnm** [comment]
I am marking this as draft to avoid accidental merge, but this should be ready for review.

**Comment by zasdfgbnm** [comment]
!build

File: csrc/executor_utils.cpp [hunk 1]
```diff
@@ -912,8 +926,10 @@ ExpressionEvaluator bindInputs(
     bool check_consistency) {
   FUSER_PERF_SCOPE("executor_utils::bindInputs");
 
+  // args may contains more than just inputs, but inputs are always at the
+  // beginning.
   TORCH_INTERNAL_ASSERT(
-      kernel->inputs().size() == args.size(),
+      kernel->inputs().size() <= args.size(),
```
Comment (jjsjann123) [inline csrc/executor_utils.cpp:932]: nitpick. maybe update the assert message.

File: test/test_allocation_domain.cpp [hunk 2]
```diff
@@ -303,4 +303,306 @@ TEST_F(AllocationDomainTest, Tensor3d_To_NHWC4d_FwdBwd_CUDA) {
...
+  tv1->axis(1)->parallelize(ParallelType::TIDx);
+  tv1->axis(0)->parallelize(ParallelType::BIDx);
+  // [BIDx, TIDx, V]
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+
+  int n = 31, h = 64, w = 103, c = 21;
+
+  at::Tensor t0_wrong_format = at::randn({n, c, h, w}, options);
+  at::Tensor t0 = t0_wrong_format.contiguous(at::MemoryFormat::ChannelsLast);
```
Comment (jjsjann123) [inline test/test_allocation_domain.cpp:344]: nitpick. maybe it's me being paranoid. But I'm pretty uncomfortable with pytorch's implicit memory_format conversion. Can we use explicit calls like `as_strided` instead? so it's easier to read the examples.

File: test/test_allocation_domain.cpp [hunk 3]
```diff
@@ -303,4 +303,306 @@ TEST_F(AllocationDomainTest, Tensor3d_To_NHWC4d_FwdBwd_CUDA) {
       __FILE__);
 }
 
+// A global->global copy kernel where both inputs are NHWC memory format
```
Comment (jjsjann123) [inline test/test_allocation_domain.cpp:306]: I think you mean `both input and output`

File: test/test_allocation_domain.cpp [hunk 4]
```diff
@@ -303,4 +303,306 @@ TEST_F(AllocationDomainTest, Tensor3d_To_NHWC4d_FwdBwd_CUDA) {
...
+  auto tv0 = makeContigConcreteTensor({n * h * w * c});
+  fusion.addInput(tv0);
+
+  std::vector<IterDomain*> tv0_1d = {tv0->axis(0)};
+  tv0->setAllocationDomain(tv0_1d, true);
+  tv0->split(0, c);
+  tv0->split(0, w);
+  tv0->split(0, h);
+  tv0->reorder({{-1, 1}});
+  tv0->commitLeafToRFactor();
```
Comment (jjsjann123) [inline test/test_allocation_domain.cpp:379]: so `commitLeafToRFactor` mutates the semantics of the input tensor -> That's why we now have input tensor represented as a 4d channels last tensor?

File: test/test_allocation_domain.cpp [hunk 5]
```diff
@@ -303,4 +303,306 @@ TEST_F(AllocationDomainTest, Tensor3d_To_NHWC4d_FwdBwd_CUDA) {
...
+
+  auto tv0 = makeContigConcreteTensor({n * h * w * c});
+  fusion.addInput(tv0);
+
+  std::vector<IterDomain*> tv0_1d = {tv0->axis(0)};
+  tv0->setAllocationDomain(tv0_1d, true);
+  tv0->split(0, c);
+  tv0->split(0, w);
+  tv0->split(0, h);
+  tv0->reorder({{-1, 1}});
```
Comment (jjsjann123) [inline test/test_allocation_domain.cpp:378]: our `reorder` always gets me by surprise. :face_with_spiral_eyes:

File: test/test_allocation_domain.cpp [hunk 6]
```diff
@@ -303,4 +303,306 @@ TEST_F(AllocationDomainTest, Tensor3d_To_NHWC4d_FwdBwd_CUDA) {
...
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+
+  at::Tensor t0_wrong_format = at::randn({n, c, h, w}, options);
+  at::Tensor t0 = t0_wrong_format.contiguous(at::MemoryFormat::ChannelsLast);
+
+  FusionExecutor fe;
+  fe.compileFusion(fusion_ptr.get(), {t0});
+
+  // EXPECT_THAT(
```
Comment (jjsjann123) [inline test/test_allocation_domain.cpp:465]: why is this commented out?

File: test/test_allocation_domain.cpp [hunk 7]
```diff
@@ -303,4 +303,306 @@ TEST_F(AllocationDomainTest, Tensor3d_To_NHWC4d_FwdBwd_CUDA) {
...
+  tv0->reorder({{-1, 1}});
+  // [N, C, H, W]
+  tv0->commitLeafToRFactor();
+
+  auto tv1 = set(tv0);
+  fusion.addOutput(tv1);
+
+  std::vector<IterDomain*> tv0_nhwc = {
+      tv0->axis(0), tv0->axis(2), tv0->axis(3), tv0->axis(1)};
+  tv0->setAllocationDomain(tv0_nhwc, true);
```
Comment (jjsjann123) [inline test/test_allocation_domain.cpp:566]: you are losing me here...

Wasn't the NHWC definition already done at the `tv0->commitLeafToRFactor()` above? what's the setAllocationDomain on tv0 doing again?!

Comparing this test with `NHWC1d_To_NHWC4d_CUDA` above, why's `setAllocationDomain` needed here?

Comment (jjsjann123) [reply test/test_allocation_domain.cpp:566]: I realize that this is likely my lack of understanding on the relationship between leaf domain / rfactor domain / allocation domain. So is there a good doc somewhere that roughly summarizes those?

File: test/test_allocation_domain.cpp (hunk 2 above)
Comment (zasdfgbnm) [reply test/test_allocation_domain.cpp:344]: Why `as_strided`? If you don't like `contiguous`, then isn't the alternative `permute` instead of `as_strided`?

Comment (jjsjann123) [reply test/test_allocation_domain.cpp:344]: lol, not arguing what's the proper way to construct inputs here. Feel free to go with what you think fits here. (even `contiguous` is probably fine, as the API is somewhat stable, but might catch you by surprise when you have broadcast dimensions - i.e. size-1 - in inputs).

I guess `as_strided` feels a lot more straightforward to me. i.e. you are providing input sizes along with their strides, so there's nothing like `permute` which actually moves dimensions around and changes their semantic order. :man_shrugging:

File: test/test_allocation_domain.cpp (hunk 7 above)
Comment (zasdfgbnm) [reply test/test_allocation_domain.cpp:566]: Unfortunately I don't think we have doc for them. We don't have a doc describing what is rFactor domain, do we? If we do have a comment in the code, I would be happy to add allocation domain to that comment.

I suggest to just think of `commitLeafToRFactor` as a convenient method for defining a tensor. For example, if we want to define a tensor
```
root: [I1, I2, I3]
rFactor: [I3, I1*I2]
```
We can of course define it manually by creating all the `IterDomain*`, `Merge*` for the root and rFactor domain, then manually create `TensorDomain*` and `TensorView*`, but this is not convenient. It is more convenient to just use our already-built utilities like `TensorView::merge`, `TensorView::reorder`, and the `commitLeafToRFactor` is exactly for that purpose.

`commitLeafToRFactor` has nothing to do with memory format, and calling `commitLeafToRFactor` means nothing about memory format. Defining the rFactor domain of a tensor using `commitLeafToRFactor` is nothing different from the plain old way of defining a tensor with rFactor domain. This method only start being heavily used in allocation domain PRs does not mean it is a utility for allocation domain. It is just because this utility is super new, and it has not been used widely yet.

The recommended mindset for reasoning about `commitLeafToRFactor` should not be "`commitLeafToRFactor` changes the rFactor domain from XXX to XXX", because this would make it confusing. The mindset I am expecting people to read codes with `commitLeafToRFactor` is, when you see code like:
```C++
auto tv = makeContigTensor(4);
tv->....;
tv->....;
tv->....;
tv->commitLeafToRFactor();
```
you will know that the code block between creating `tv` and `commitLeafToRFactor` are the region that you define this tensor. So you should automatically add some comment in your mind
```C++
auto tv = makeContigTensor(4);  // begin defining tv. From now on, I will build a graph of IterDomains connecting root and rFactor domain
tv->....;
tv->....;
tv->....;
tv->commitLeafToRFactor();  // end defining tv. OK, the rFactor domain is built. The relationship between root and rFactor domain is blabla, so this is a view+permute op that changes the tensor semantics from XXX to YYY.
```

Comment (zasdfgbnm) [reply test/test_allocation_domain.cpp:566]: Whether a tensor is NHWC or NCHW depend on whether the order of `IterDomain`s of rFactor domain and allocation domain match. The key is whether rFactor match with allocation, and it has nothing to do with the relationship between rFactor domain and root domain.

Comment (zasdfgbnm) [reply test/test_allocation_domain.cpp:566]: The rFactor domain should always be `NCHW`, regardless of the memory format, and if allocation domain match with rFactor domain, then this tensor's memory format is also `NCHW`. But if the order of allocation domain is `[N, H, W, C]`, or more complicated `[N*H*W*C]`, `[N*H, W*C]`, `[N, H*W, C]`, `[N, H, W, C/5, 5]`, then this tensor a NHWC tensor.

File: test/test_allocation_domain.cpp (hunk 2 above)
Comment (zasdfgbnm) [reply test/test_allocation_domain.cpp:344]: Wait, no, I think neither `contiguous` nor `permute` works. It has to be `contiguous`. Because I want to keep the semantics of the tensor, but store it differently. Neither `as_strided` nor `permute` change the underlying tensor storage.

File: test/test_allocation_domain.cpp (hunk 7 above)
Comment (zasdfgbnm) [reply test/test_allocation_domain.cpp:566]: Also, rFactor domain only has something to do with tensor semantics, it has nothing to do with memory format. This is just like in PyTorch, having a tensor with semantics `NCHW` does not means it is a `NCHW` tensor, and doing a permute on that tensor to change its semantics `NHWC` does not mean you changed the memory format from `NCHW` to `NHWC`.

File: test/test_allocation_domain.cpp (hunk 2 above)
Comment (jjsjann123) [reply test/test_allocation_domain.cpp:344]: > Wait, no, I think neither `as_strided`  nor `permute` works. It has to be `contiguous`.
^^^ I don't agree with this.
I agree that `contiguous` is different from the other two and they can't be used interchangeably. But in our test, we don't care about the exact data, since we control the reference as well.

i.e. we can create a contiguous buffer and call `as_strided` to create a tensor with a semantic size and an arbitrary memory storage. As long as: 1. it matches fusion definition; 2. the same tensor is used for reference computation.

Comment (jjsjann123) [reply test/test_allocation_domain.cpp:344]: similarly, in order to create a [2, 3, 4, 5] tensor in channels-last format, you can create a [2, 4, 5, 3] tensor an permute it to get a size [2, 3, 4, 5] in channels-last format.
It's not going to be the same tensor, but it doesn't matter for our tests.

File: test/test_allocation_domain.cpp (hunk 7 above)
Comment (jjsjann123) [reply test/test_allocation_domain.cpp:566]: Thanks a lot. This sounds like what I remembered in our previous conversations, but having it written down here is helpful.

I didn't realize that we can have view/permute setup directly on input tensor, but I guess that makes sense.

One confusion is still bothering me even after the read.
```
 std::vector<IterDomain*> tv0_2d = {tv0->axis(0), tv0->axis(1)};
  tv0->setAllocationDomain(tv0_2d, true);
  tv0->merge(0);
  tv0->split(0, c);
  tv0->split(0, w);
  tv0->split(0, h);
  // [N, H, W, C]
  tv0->reorder({{-1, 1}});
  // [N, C, H, W]
  tv0->commitLeafToRFactor();

  auto tv1 = set(tv0);
  fusion.addOutput(tv1);

  std::vector<IterDomain*> tv0_nhwc = {
      tv0->axis(0), tv0->axis(2), tv0->axis(3), tv0->axis(1)};
  tv0->setAllocationDomain(tv0_nhwc, true);
```
What's the purpose of `tv0->setAllocationDomain(tv0_2d, true);`?
if all transformation done until `commitLeafToRFactor()` is only semantic-related, that allocation assign in the beginning doesn't seem to be doing anything.
And later we are setting allocation domain explicitly again. which should overwrite the first allocation setup?!

Comment (zasdfgbnm) [reply test/test_allocation_domain.cpp:566]: Oh, I see where the confusion is. I copy-pasted tests from my previous prototyping branch. I remembered that these tests were already well written and cleaned up. Seems that my memory was wrong. Sorry about the confusion. I will update the test. And yes, your understanding is correct. Setting the allocation domain twice overwrite the first result, so it makes no sense to do so.

File: csrc/dynamic_transform.cpp [hunk 8]
```diff
@@ -7,6 +7,7 @@
 // clang-format on
 #include <device_lower/utils.h>
 #include <dynamic_transform.h>
+#include <executor_kernel_arg.h>
```
Comment (naoyam) [inline csrc/dynamic_transform.cpp:10]: Why is this added?

File: csrc/executor_kernel_arg.h [hunk 9]
```diff
@@ -59,16 +60,19 @@ inline std::string argTypeToString(ArgType type) {
...
   using index_type = nvfuser_index_t;
 
   void* data;
   std::array<nvfuser_index_t, ndims> size;
-  std::array<nvfuser_index_t, ndims> stride;
+  std::array<nvfuser_index_t, nalloc> stride;
   constexpr int nDims() const {
     return ndims;
   }
+  constexpr int nAllocationDims() const {
```
Comment (naoyam) [inline csrc/executor_kernel_arg.h:73]: I guess it doesn't matter as it's `constexpr`, but this should be `static`

File: test/test_allocation_domain.cpp (hunk 2 above)
Comment (zasdfgbnm) [reply test/test_allocation_domain.cpp]: I updated the test to use `as_strided`. In terms of interface, I definitely prefer the `contiguous(at::MemoryFormat::ChannelsLast)` one. It is very unfortunate that this interface is used for an unpredictable and confusing behavior.

File: test/test_allocation_domain.cpp (hunk 3 above)
Comment (zasdfgbnm) [reply test/test_allocation_domain.cpp]: Fixed, thanks for catching!

File: csrc/executor_kernel_arg.cpp [hunk 10]
```diff
@@ -14,6 +14,292 @@
...
+    // TORCH_INTERNAL_ASSERT(in_it != active_ids_.end())
+    if (in_it == active_ids_.end()) {
+      // TODO: see [Allocation domain on both side of rFactor]
+      return;
+    }
+    auto [in_size, in_stride] = in_it->second;
+    auto factor = ee_.evaluate(split->factor())->as<int64_t>();
+    TORCH_INTERNAL_ASSERT(
+        in_size % factor == 0,
+        "non-divisible split is not allowed in allocation domain");
```
Comment (naoyam) [inline csrc/executor_kernel_arg.cpp:40]: Why isn't it allowed?

File: csrc/executor_kernel_arg.cpp [hunk 11]
```diff
@@ -14,6 +14,292 @@
...
+    // TORCH_INTERNAL_ASSERT(outer_it != active_ids_.end())
+    if (inner_it == active_ids_.end() || outer_it == active_ids_.end()) {
+      // TODO: see [Allocation domain on both side of rFactor]
+      return;
+    }
+    auto [inner_size, inner_stride] = inner_it->second;
+    auto [outer_size, outer_stride] = outer_it->second;
+    TORCH_INTERNAL_ASSERT(
+        inner_stride * inner_size == outer_stride,
+        "Merging of discontiguous dimensions is not allowed in allocation domain");
```
Comment (naoyam) [inline csrc/executor_kernel_arg.cpp:70]: Why isn't it allowed?

File: csrc/executor.cpp [hunk 12]
```diff
@@ -1009,6 +1009,8 @@ std::vector<at::Tensor> allocOutputs(
           at::TensorOptions().dtype(at::kFloat).device(device);
       outputs.emplace_back(
           at::empty(std::vector<int64_t>(alloc_dom.size(), 0), tensor_options));
+      outputs.emplace_back(
```
[the remark this answers is not in the document]
Comment (zasdfgbnm) [reply csrc/executor.cpp:1012]: This definitely is a duplicate. Looks like this comes from merge conflict not correctly resolved, and I pushed before running all tests. Will fix.

File: csrc/executor_utils.cpp (hunk 1 above)
Comment (zasdfgbnm) [reply csrc/executor_utils.cpp:932]: updated

File: test/test_allocation_domain.cpp (hunk 4 above)
Comment (zasdfgbnm) [reply test/test_allocation_domain.cpp:380]: Yes, for this example, you allocate as a 1d flatten tensor, and view that allocation as a [H, C, H, W] semantics

File: test/test_allocation_domain.cpp (hunk 6 above)
Comment (zasdfgbnm) [reply test/test_allocation_domain.cpp]: I commented this out while debugging, and forget to uncomment it. Fixed.

File: csrc/dynamic_transform.cpp (hunk 8 above)
Comment (zasdfgbnm) [reply csrc/dynamic_transform.cpp:10]: This file uses `KernelArgumentHolder` but didn't include its header. In the past, this was not failing because it included `evaluator_common.h` which then included `executor_kernel_arg.h`, but now, `evaluator_common.h` is no longer including `executor_kernel_arg.h`. In general, a file should include everything it uses. It should not rely on other files to include things for it.

File: csrc/executor_kernel_arg.h (hunk 9 above)
Comment (zasdfgbnm) [reply csrc/executor_kernel_arg.h]: fixed

File: test/test_allocation_domain.cpp (hunk 7 above)
Comment (zasdfgbnm) [reply test/test_allocation_domain.cpp]: Updated the test

File: csrc/executor_kernel_arg.cpp (hunk 10 above)
Comment (zasdfgbnm) [reply csrc/executor_kernel_arg.cpp:40]: This analysis will only run on tensors provided by PyTorch. PyTorch does not allow, for example, viewing a tensor of shape  `(1, 5)` as `(4,)`.

File: csrc/executor_kernel_arg.cpp (hunk 11 above)
Comment (zasdfgbnm) [reply csrc/executor_kernel_arg.cpp:70]: If you merge two IDs into one, these two IDs must be contiguous. This is just like, if you have a tensor whose shape is `(3, 4)`, you can view it as `(12,)` only if it is contiguous. Otherwise 
```
for i in range(12):
  index = i * stride[0]
```
will fail.

Comment (zasdfgbnm) [reply csrc/executor_kernel_arg.cpp:70]: Maybe a better way to explain this is, if you have a tensor of shape `(3, 4)`, you can allocate a flat size 12 tensor to store this `(3, 4)` tensor if and only if the `3` is contiguous w.r.t. the `4`.

**Comment by zasdfgbnm** [comment]
!build

File: test/test_allocation_domain.cpp (hunk 7 above)
Comment (zasdfgbnm) [reply test/test_allocation_domain.cpp]: Updated the test

File: csrc/executor_kernel_arg.cpp (hunk 10 above)
Comment (naoyam) [reply csrc/executor_kernel_arg.cpp:40]: I see. Can you add that to the error message? At first, I thought it's not allowed in nvfuser in general, which is not true.

File: csrc/executor_kernel_arg.cpp (hunk 11 above)
Comment (naoyam) [reply csrc/executor_kernel_arg.cpp:70]: Makes sense.

File: csrc/executor_kernel_arg.cpp [hunk 13]
```diff
@@ -14,6 +14,292 @@
...
+  // Validate final sizes and strides with contiguity
+  int64_t contiguous_stride = 1;
+  std::vector<std::optional<bool>> contiguity = tv->getContiguity();
+  for (int64_t i = (int64_t)sizes_strides.size() - 1; i >= 0; i--) {
+    if (alloc.at(i)->isBroadcast()) {
+      continue;
+    }
+    while (!contiguity.back().has_value()) {
+      contiguity.pop_back();
+      continue;
```
Comment (naoyam) [inline csrc/executor_kernel_arg.cpp:262]: This `continue` seems redundant.

File: csrc/executor_kernel_arg.cpp [hunk 14]
```diff
@@ -14,6 +14,292 @@
...
+  for (auto i : c10::irange(alloc.size())) {
+    auto id = alloc.at(i);
+    sizes_strides.emplace_back(active_ids.at(id));
+  }
+  // Validate final sizes and strides with contiguity
+  int64_t contiguous_stride = 1;
+  std::vector<std::optional<bool>> contiguity = tv->getContiguity();
+  for (int64_t i = (int64_t)sizes_strides.size() - 1; i >= 0; i--) {
+    if (alloc.at(i)->isBroadcast()) {
+      continue;
```
Comment (naoyam) [inline csrc/executor_kernel_arg.cpp:258]: Shouldn't `conitiguity` be popped here too?

File: csrc/executor_kernel_arg.cpp (hunk 13 above)
Comment (naoyam) [reply csrc/executor_kernel_arg.cpp:262]: Actually, not really sure what this while loop is trying to do.

File: csrc/executor_kernel_arg.cpp [hunk 15]
```diff
@@ -14,6 +14,292 @@
...
+    }
+    while (!contiguity.back().has_value()) {
+      contiguity.pop_back();
+      continue;
+    }
+    auto [size, stride] = sizes_strides.at(i);
+    auto last_contiguity = contiguity.back();
```
Comment (naoyam) [inline csrc/executor_kernel_arg.cpp:265]: Do we want to assert `contiguity` is not empty here?

File: csrc/executor_kernel_arg.cpp [hunk 16]
```diff
@@ -24,42 +310,94 @@ PrimDataType TensorArgAbstract::getSmallestIndexType() const {
     }
   }
   return PrimDataType::Int32;
+  return PrimDataType::Int32;
```
Comment (naoyam) [inline csrc/executor_kernel_arg.cpp:313]: I guess this is due to a merge. Wonder why clang-tidy didn't say anything.

File: test/test_allocation_domain.cpp [hunk 17]
```diff
@@ -303,4 +303,307 @@ TEST_F(AllocationDomainTest, Tensor3d_To_NHWC4d_FwdBwd_CUDA) {
...
+
+  // [N, C, H, W]
+  tv1->reorder({{1, -1}});
+  // [N, H, W, C]
+  tv1->merge(0);
+  tv1->merge(0);
+  tv1->merge(0);
+  // [N*H*W*C]
+  tv1->split(0, 4);
+  tv1->axis(1)->parallelize(ParallelType::Vectorize);
```
Comment (naoyam) [inline test/test_allocation_domain.cpp:334]: Nice!

File: csrc/executor_kernel_arg.cpp (hunk 10 above)
Comment (zasdfgbnm) [reply csrc/executor_kernel_arg.cpp:42]: Added

File: csrc/executor_kernel_arg.cpp (hunk 15 above)
Comment (zasdfgbnm) [reply csrc/executor_kernel_arg.cpp:273]: Added

File: csrc/executor_kernel_arg.cpp (hunk 16 above)
Comment (zasdfgbnm) [reply csrc/executor_kernel_arg.cpp]: fixed

File: csrc/executor_kernel_arg.cpp (hunk 13 above)
Comment (zasdfgbnm) [reply csrc/executor_kernel_arg.cpp]: `continue` removed. The size of `contiguity` does not necessarily equal to the size of `sizes_strides`, because reductions do have contiguity but are not in `sizes_strides`. So this loop just pop all null contiguity on the back, these null contiguity might be from a reduction or from a broadcast of the last iteration of the outer loop. Whatever this null contiguity comes from, it does not belong to this iteration, so it should be just skipped.

File: csrc/executor_kernel_arg.cpp (hunk 14 above)
Comment (zasdfgbnm) [reply csrc/executor_kernel_arg.cpp:266]: No, see https://github.com/NVIDIA/Fuser/pull/338#discussion_r1195874473

File: csrc/executor_kernel_arg.cpp (hunk 13 above)
Comment (zasdfgbnm) [reply csrc/executor_kernel_arg.cpp]: I added a `TORCH_INTERNAL_ASSERT(contiguity.empty(),` to then end of the outer loop to check if the "actual size" of `contiguity` matches with `sizes_strides`.

**Comment by zasdfgbnm** [comment]
!build

**Review by naoyam — APPROVED** [review]
LGTM
